src/hooks.rs

```diff
@@ -36,365 +31,555 @@ impl From<i32> for Action {
-        } else {
-            None
-        };
-
-        let previous_hook = match hook {
-            Some(hook) => {
-                let boxed_hook: *mut F = Box::into_raw(Box::new(hook));
-                unsafe {
-                    ffi::sqlite3_commit_hook(
-                        self.db(),
-                        Some(call_boxed_closure::<F>),
-                        boxed_hook as *mut _,
-                    )
+            let free_preupdate_hook = if hook.is_some() {
+                Some(free_boxed_hook::<F> as unsafe fn(*mut c_void))
+            } else {
+                None
+            };
+
+            let previous_hook = match hook {
+                Some(hook) => {
+                    let boxed_hook: *mut F = Box::into_raw(Box::new(hook));
+                    unsafe {
+                        ffi::sqlite3_preupdate_hook(
+                            self.db(),
+                            Some(call_boxed_closure::<F>),
+                            boxed_hook as *mut _,
+                        )
+                    }
+                }
+                _ => unsafe { ffi::sqlite3_preupdate_hook(self.db(), None, ptr::null_mut()) },
+            };
+            if !previous_hook.is_null() {
+                if let Some(free_boxed_hook) = self.free_preupdate_hook {
+                    unsafe { free_boxed_hook(previous_hook) };
                 }
             }
-            _ => unsafe { ffi::sqlite3_commit_hook(self.db(), None, ptr::null_mut()) },
-        };
-        if !previous_hook.is_null() {
-            if let Some(free_boxed_hook) = self.free_commit_hook {
-                unsafe { free_boxed_hook(previous_hook) };
-            }
+            self.free_preupdate_hook = free_preupdate_hook;
+        }
+    }
+
+    #[cfg(test)]
+    mod test {
+        use super::super::Action;
+        use super::PreUpdateHookFunctions;
+        use crate::{Connection, Result};
+
+        #[test]
+        fn test_preupdate_hook() -> Result<()> {
+            let db = Connection::open_in_memory()?;
+
+            let mut called = false;
+            db.preupdate_hook(Some(
+                |action,
+                 db: &str,
+                 tbl: &str,
+                 row_id,
+                 new_row_id,
+                 _func: &PreUpdateHookFunctions| {
+                    assert_eq!(Action::SQLITE_INSERT, action);
+                    assert_eq!("main", db);
+                    assert_eq!("foo", tbl);
+                    assert_eq!(1, row_id);
+                    assert_eq!(1, new_row_id);
+                    called = true;
+                },
+            ));
+            db.execute_batch("CREATE TABLE foo (t TEXT)")?;
+            db.execute_batch("INSERT INTO foo VALUES ('lisa')")?;
+            assert!(called);
+            Ok(())
         }
-        self.free_commit_hook = free_commit_hook;
     }
+}
+
+#[cfg(feature = "hooks")]
+mod datachanged_and_friends {
```

Comment:
This is just a preliminary name, I couldn't come up with a succint naming for all the hooks, excluding the preupdate_hook.